extra_foam/gui/image_tool/reference_view.py

```diff
@@ -76,26 +103,107 @@ def _loadReference(self):
 
     @pyqtSlot()
-    def _setReference(self):
+    def _setCurrentAsReference(self):
         """Set the current corrected image as reference."""
-        img = self._corrected.image
-        if img is not None:
-            filepath = osp.join(ROOT_PATH, "tmp", ".reference.npy")
-            if not osp.exists(osp.dirname(filepath)):
-                os.mkdir(osp.dirname(filepath))
-
-            try:
-                write_image(filepath, img)
-            except ValueError as e:
-                logger.error(str(e))
-
-            self._pub.set(filepath)
-            self._ctrl_widget.filepath_le.setText(filepath)
+        image = self._corrected.image
+        if image is not None:
+            self._write_temp_reference_file(image)
+            if self._check_reference_file():
+                self._set_reference(REFERENCE_FILE)
+                # Update label
+                text = "The current image has been applied as reference."
+                self._ctrl_widget.record_label.setText(text)
+
+    def _write_temp_reference_file(self, image):
+        """Save the input image as a reference file (.npy).
+           Path is $EXTRA_FOAM_TEMP_FOLDER/.reference.npy"""
+        if image is None:
+            return
+
+        os.makedirs(osp.dirname(REFERENCE_FILE), exist_ok=True)
+
+        try:
+            write_image(REFERENCE_FILE, image)
+        except ValueError as e:
+            logger.error(str(e))
+            # Remove previous reference file
+            os.remove(REFERENCE_FILE)
+
+    def _set_reference(self, path):
+        """Sets the input image as reference"""
+        self._pub.set(path)
+        self._ctrl_widget.filepath_le.setText(path)
 
     @pyqtSlot()
     def _removeReference(self):
         """Remove the reference image."""
-        self._pub.set("")
-        self._ctrl_widget.filepath_le.setText("")
+        self._set_reference("")
+        self._ctrl_widget.record_label.setText("")
+
+    @pyqtSlot(bool)
+    def _recordReference(self, is_recording):
+        """Start/stop the recording of the reference file.
+
+           When starting, a flag will be set for the `updateF()` to detect the
+           start of the image averaging of the received data.
+           Upon finishing, the recorded image is written as reference file.
+        """
+        self._is_recording = is_recording
+
+        # Log recording status
+        status = "started" if is_recording else "finished"
+        logger.info(f"Recording reference has {status}.")
+
+        if is_recording:
+            # Remove existing reference file
+            self._removeReference()
+        elif self._recorded_image is not None:
+            # Save recording to a temp file upon finishing
+            self._write_temp_reference_file(self._recorded_image)
+            if self._check_reference_file():
+                self._set_reference(REFERENCE_FILE)
+
+            # Reset widget and variables
+            self._check_reference_file()
```

Comment:
Is this second call necessary since `self._check_reference_file()` is already called in the if-statement above?

Reply:
Actually, no. IIRC these were two different logic before I merged them into one XD Removed.